[technotes] naming conventions (#117)
apollographql/docs: src/content/technotes/TN0002-naming-conventions.mdx

---

> title: TN0002 - Naming Conventions

Reviewer: ```suggestion
title: TN0002 - Naming conventions
```

> title: TN0002 - Naming conventions

Reviewer 2: ```suggestion
title: TN0002 - GraphQL naming conventions
```

Reviewer 2: Right now this tech note doesn't cover naming _operations_. It's probably worth adding a short section for that, since there isn't a lot to it and probably doesn't merit its own separate tech note in the future.

I think the guidance we provide elsewhere in the docs and Odyssey amounts to: PascalCase + start with a verb (usually `Get` for a query, and a relevant verb for the action of a mutation)

Author: How about "Schema naming conventions"? I want this tagged as "Schema design" and intended for schema developers, and we can have a separate note for frontend devs.

Reviewer 2: Sure!

> description: Naming conventions for types, fields, and arguments.

Reviewer 2: ```suggestion
```

Reviewer 2: (I feel like the title is sufficiently descriptive in this case)

Author: what is the title used for? i assumed it's really just for search results, so i figured having "types" "fields" and "arguments" might help. but i'm fine dropping it

Reviewer 2: Yeah I guess it's best to leave this for SEO, I think it gets included in Google results 🤔 

Let's change it to "Conventions for types, fields, and arguments" (cut "Naming" and the trailing period)

tags: [schema design]
---

> ## General Best Practices

Reviewer 2: ```suggestion
## High-level guidance
```

- Regardless of your chosen conventions, be consistent across the entire schema.

> - Be specific; don't "land grab" names with broad applicability.

Reviewer 2: ```suggestion
- Be specific with names—don't "land grab" names with broad applicability.
```

- Avoid acronyms, initialisms, and abbreviations.

### Casing

> Use `camelCase` for field names, argument names, and directive names.

Reviewer 2: ```suggestion
Use `camelCase` for field names, argument names, and directive names:
```

Reviewer 2: Just a universal thing I do, which is when a sentence leads into a codeblock example, end it with a colon

```graphql {2,5}
type Query {
  myCamelCaseFieldNames(myArgumentName: String): String
}

directive @myDirective on FIELD
```

> Use `PascalCase` or `UpperCamelCase` for type names.

Reviewer 2: ```suggestion
Use `PascalCase` for type names:
```

Author: I was referencing a few different guides which used different names for this. But 👍🏻

```graphql
type MyType { ... }

enum MyEnum { ... }

interface MyInterface { ... }

union MyUnion = ...

scalar MyScalar
```

> Use `SCREAMING_SNAKE_CASE` for enum values.

Reviewer 2: ```suggestion
Use `SCREAMING_SNAKE_CASE` for enum values:
```

Reviewer 2: Also 😆 never seen that term before

```graphql {2,3}
enum MyEnum {
  VALUE_ONE
  VALUE_TWO
}
```

> ### Field Names

Reviewer 2: ```suggestion
### Field names
```

Avoid verb prefixes like `get` or `list` on query (read) fields.

```graphql {6}
type Query {
  # ❌ incorrect
  getProducts: [Product]

  # ✅ correct
  products: [Product]
}
```

> This creates symmetry between root fields and nested fields:

Reviewer 2: ```suggestion
This creates consistency between root fields and nested fields:
```

```graphql
# ❌ incorrect
query Products {
  getProducts {
    id
    getReviews {
      content
    }
  }
}

# ✅ correct
query Products {
  products {
    id
    reviews {
      content
    }
  }
}
```

> Start mutation fields with a verb.

Reviewer 2: ```suggestion
Start mutation fields with a verb:
```

```graphql {6}
type Mutation {
  # ❌ incorrect
  customerAdd(input: AddCustomerInput): AddCustomerPayload!

  # ✅ correct
  addCustomer(input: AddCustomerInput): AddCustomerPayload!
}
```

> ### Type Names

Reviewer 2: ```suggestion
### Type names
```

> Use the suffix `Input` when naming input types.

Reviewer 2: ```suggestion
Use the suffix `Input` when naming input types:
```

```graphql {1}
input AddCustomerInput {
  name: String!
}
```

> Use the suffix `Payload` when naming output types returned from mutations.

Reviewer 2: In the docs, we recommend the suffix `Response`: https://www.apollographql.com/docs/apollo-server/schema/schema#structuring-mutation-responses

Can we modify this guidance to say "Use a consistent suffix like `Response` or `Payload` when..."

Author: excellent change

```graphql {5}
type Mutation {
  addCustomer(input: AddCustomerInput!): AddCustomerPayload!
}

type AddCustomerPayload {
  success: Boolean!
}
```

> ## Additional Considerations

Reviewer 2: ```suggestion
## Additional considerations
```

> ### Enforcement

Reviewer 2: ```suggestion
### Enforcing conventions
```

Use GraphQL-ESLint's [naming-convention rule](https://github.com/B2o5T/graphql-eslint/blob/master/docs/rules/naming-convention.md) to catch violations.

### Namespacing

> When resolving naming conflicts between different domains, choose between these two conventions:

Reviewer 2: ```suggestion
When resolving naming conflicts between different domains, we recommend using one of the following conventions:
```
